File: app/src/main/java/com/dcrandroid/activities/SplashScreen.java
/*
 * Copyright (c) 2018-2019 The Decred developers
 * Use of this source code is governed by an ISC
 * license that can be found in the LICENSE file.
 */

package com.dcrandroid.activities;

import android.content.DialogInterface;
import android.content.Intent;
import android.graphics.drawable.Drawable;
import android.os.Bundle;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import androidx.core.app.ActivityCompat;
import androidx.vectordrawable.graphics.drawable.Animatable2Compat;
import androidx.vectordrawable.graphics.drawable.AnimatedVectorDrawableCompat;

import com.dcrandroid.BuildConfig;
import com.dcrandroid.HomeActivity;
import com.dcrandroid.MainActivity;
import com.dcrandroid.R;
import com.dcrandroid.data.Constants;
import com.dcrandroid.dialog.InfoDialog;
import com.dcrandroid.util.PreferenceUtil;
import com.dcrandroid.util.Utils;
import com.dcrandroid.util.WalletData;

import java.io.File;

import dcrlibwallet.Dcrlibwallet;
import dcrlibwallet.LibWallet;

/**
 * Created by Macsleven on 24/12/2017.
 */

public class SplashScreen extends BaseActivity {

    private final int PASSWORD_REQUEST_CODE = 1;
    private ImageView imgAnim;
    private PreferenceUtil util;
    private TextView tvLoading;
    private Thread loadThread;
    private WalletData walletData;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        if (!isTaskRoot()) {
            final Intent intent = getIntent();
            final String intentAction = intent.getAction();
            if (intent.hasCategory(Intent.CATEGORY_LAUNCHER) && intentAction != null && intentAction.equals(Intent.ACTION_MAIN)) {
                finish();
                return;
            }
        }

        util = new PreferenceUtil(this);
        setContentView(R.layout.splash_page);

        if (BuildConfig.IS_TESTNET) {
            findViewById(R.id.tv_testnet).setVisibility(View.VISIBLE);
        }

        imgAnim = findViewById(R.id.splashscreen_icon);
        imgAnim.setOnClickListener(new DoubleClickListener() {
            @Override
            public void onSingleClick(View v) {

            }

            @Override
            public void onDoubleClick(View v) {
                if (loadThread != null) {
                    loadThread.interrupt();
                }

                if (walletData.wallet != null) {
                    walletData.wallet.shutdown();
                }

                Intent intent = new Intent(getApplicationContext(), SettingsActivity.class);
                startActivityForResult(intent, 2);
            }
        });

        imgAnim.post(new Runnable() {
            @Override
            public void run() {

                final AnimatedVectorDrawableCompat anim = AnimatedVectorDrawableCompat.create(getApplicationContext(), R.drawable.avd_anim);
                anim.registerAnimationCallback(new Animatable2Compat.AnimationCallback() {
                    @Override
                    public void onAnimationEnd(Drawable drawable) {
                        super.onAnimationEnd(drawable);
                        imgAnim.postDelayed(new Runnable() {
                            @Override
                            public void run() {
                                anim.start();
                            }
                        }, 750);
                    }
                });
                imgAnim.setImageDrawable(anim);
                anim.start();
            }
        });

        tvLoading = findViewById(R.id.loading_status);
        startup();
    }

    private void startup() {
        walletData = WalletData.getInstance();

        try {
            if (walletData.wallet != null && walletData.wallet.walletOpened()) {
                walletData.wallet.shutdown();
            }

            walletData.wallet = null;
        } catch (Exception e) {
            e.printStackTrace();
        }

        String homeDir = getFilesDir() + "/wallet";
        walletData.wallet = new LibWallet(homeDir, Constants.BADGER_DB, BuildConfig.NetType);
        Dcrlibwallet.setLogLevels(util.get(Constants.LOGGING_LEVEL));

        String walletDB;

        if (BuildConfig.IS_TESTNET) {
            walletDB = "/testnet3/wallet.db";
        } else {
            walletDB = "/mainnet/wallet.db";
        }

        File f = new File(homeDir, walletDB);
        if (!f.exists()) {
            loadThread = new Thread() {
                public void run() {
                    try {
                        sleep(3000);
                        createWallet();
                    } catch (InterruptedException e) {
                        e.printStackTrace();
                    }
                }
            };
            loadThread.start();
        } else {
            checkEncryption();
        }
    }

    private void setText(final String str) {
        runOnUiThread(new Runnable() {
            @Override
            public void run() {
                tvLoading.setText(str);
            }
        });
    }

    private void createWallet() {
        Intent i = new Intent(SplashScreen.this, SetupWalletActivity.class);
        startActivity(i);
        finish();
    }

    private void openWallet(final String publicPass) {
        loadThread = new Thread() {
            public void run() {
                try {
                    setText(getString(R.string.opening_wallet));
                    walletData.wallet.openWallet(publicPass.getBytes());
                    Intent i = new Intent(SplashScreen.this, HomeActivity.class);
                    i.addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP);
                    startActivity(i);
                    //Finish all the activities before this
                    ActivityCompat.finishAffinity(SplashScreen.this);
                } catch (final Exception e) {
                    e.printStackTrace();
                    runOnUiThread(new Runnable() {
                        @Override
                        public void run() {
                            InfoDialog infoDialog = new InfoDialog(SplashScreen.this)
                                    .setDialogTitle(getString(R.string.failed_to_open_wallet))
                                    .setMessage(Utils.translateError(SplashScreen.this, e))
                                    .setIcon(R.drawable.np_amount_withdrawal) //Temporary Icon
                                    .setPositiveButton(getString(R.string.exit_cap), new DialogInterface.OnClickListener() {
                                        @Override
                                        public void onClick(DialogInterface dialog, int which) {
                                            finish();
                                        }
                                    });

                            if (e.getMessage().equals(Dcrlibwallet.ErrInvalidPassphrase)) {
                                if (util.get(Constants.STARTUP_PASSPHRASE_TYPE).equals(Constants.PIN)) {
                                    infoDialog.setMessage(getString(R.string.invalid_pin));
                                }
                                infoDialog.setNegativeButton(getString(R.string.exit_cap), new DialogInterface.OnClickListener() {
                                    @Override
                                    public void onClick(DialogInterface dialog, int which) {
                                        finish();
                                    }
                                }).setPositiveButton(getString(R.string.retry_caps), new DialogInterface.OnClickListener() {
                                    @Override
                                    public void onClick(DialogInterface dialog, int which) {
                                        Intent i;

                                        if (util.get(Constants.STARTUP_PASSPHRASE_TYPE).equals(Constants.PASSWORD)) {
                                            i = new Intent(SplashScreen.this, EnterPasswordActivity.class);
                                        } else {
                                            i = new Intent(SplashScreen.this, EnterPassCode.class);
                                        }

                                        i.putExtra(Constants.SPENDING_PASSWORD, false);
                                        i.putExtra(Constants.NO_RETURN, true);
                                        startActivityForResult(i, PASSWORD_REQUEST_CODE);
                                    }
                                });
                            }

                            infoDialog.setCancelable(false);
                            infoDialog.setCanceledOnTouchOutside(false);
                            infoDialog.show();

                        }
                    });
                }
            }
        };
        loadThread.start();
    }

    public void checkEncryption() {
        if (util.getBoolean(Constants.ENCRYPT)) {
            Intent i;

            if (util.get(Constants.STARTUP_PASSPHRASE_TYPE).equals(Constants.PASSWORD)) {
                i = new Intent(this, EnterPasswordActivity.class);
            } else {
                i = new Intent(this, EnterPassCode.class);
            }

            i.putExtra(Constants.SPENDING_PASSWORD, false);
            i.putExtra(Constants.NO_RETURN, true);
            startActivityForResult(i, PASSWORD_REQUEST_CODE);
        } else {
            openWallet(Constants.INSECURE_PUB_PASSPHRASE);
        }
    }

    @Override
    protected void onActivityResult(int requestCode, int resultCode, Intent data) {
        super.onActivityResult(requestCode, resultCode, data);
        if (requestCode == 2) {
            startup();
        } else if (requestCode == PASSWORD_REQUEST_CODE) {
            if (resultCode == RESULT_OK) {
                openWallet(data.getStringExtra(Constants.PASSPHRASE));
            }
        }
    }

    public abstract class DoubleClickListener implements View.OnClickListener {

        private static final long DOUBLE_CLICK_TIME_DELTA = 300;//milliseconds

        long lastClickTime = 0;

        @Override
        public void onClick(View v) {
            long clickTime = System.currentTimeMillis();
            if (clickTime - lastClickTime < DOUBLE_CLICK_TIME_DELTA) {
                onDoubleClick(v);
                lastClickTime = 0;
            } else {
                onSingleClick(v);
            }
            lastClickTime = clickTime;
        }

        public abstract void onSingleClick(View v);

        public abstract void onDoubleClick(View v);
    }
}

File: app/src/main/java/com/dcrandroid/activities/Splashscreen.kt
/*
 * Copyright (c) 2018-2019 The Decred developers
 * Use of this source code is governed by an ISC
 * license that can be found in the LICENSE file.
 */

package com.dcrandroid.activities

import android.content.DialogInterface
import android.content.Intent
import android.graphics.drawable.Drawable
import android.os.Bundle

import androidx.core.app.ActivityCompat
import androidx.vectordrawable.graphics.drawable.Animatable2Compat
import androidx.vectordrawable.graphics.drawable.AnimatedVectorDrawableCompat

import com.dcrandroid.BuildConfig
import com.dcrandroid.HomeActivity
import com.dcrandroid.R
import com.dcrandroid.data.Constants
import com.dcrandroid.dialog.InfoDialog
import com.dcrandroid.extensions.show
import com.dcrandroid.util.*

import dcrlibwallet.Dcrlibwallet
import dcrlibwallet.MultiWallet
import kotlinx.android.synthetic.main.splash_page.*
import kotlinx.coroutines.*
import kotlin.system.exitProcess

/**
 * Created by Macsleven on 24/12/2017.
 */

class SplashScreen : BaseActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        if (!isTaskRoot) {
            val intent = intent
            val intentAction = intent.action
            if (intent.hasCategory(Intent.CATEGORY_LAUNCHER) && intentAction != null && intentAction == Intent.ACTION_MAIN) {
                finish()
                return
            }
        }

        setContentView(R.layout.splash_page)

        if (BuildConfig.IS_TESTNET) {
            tv_testnet.show()
        }

        splashscreen_dcr_symbol.post {
            val anim = AnimatedVectorDrawableCompat.create(applicationContext, R.drawable.avd_anim)
            anim!!.registerAnimationCallback(object : Animatable2Compat.AnimationCallback() {
                override fun onAnimationEnd(drawable: Drawable?) {
                    super.onAnimationEnd(drawable)
                    splashscreen_dcr_symbol.postDelayed({ anim.start() }, 750)
                }
            })
            splashscreen_dcr_symbol.setImageDrawable(anim)
            anim.start()
        }

        startup()
    }

    private fun setText(str: String) = GlobalScope.launch(Dispatchers.Main) {
        loading_status.text = str
    }

    private fun startup() {

        try {
            if (multiWallet != null) {
                multiWallet!!.shutdown()
            }

            walletData.multiWallet = null
        } catch (e: Exception) {
            e.printStackTrace()
        }

        val homeDir = "$filesDir/$walletsDirName"
        walletData.multiWallet = MultiWallet(homeDir, Constants.BADGER_DB, BuildConfig.NetType)

        val logLevels = resources.getStringArray(R.array.logging_levels)
        val logLevelIndex = multiWallet!!.readInt32ConfigValueForKey(Dcrlibwallet.LogLevelConfigKey, Constants.DEF_LOG_LEVEL)
        Dcrlibwallet.setLogLevels(logLevels[logLevelIndex])

        if (multiWallet!!.loadedWalletsCount() == 0) {

            val v1WalletPath = "$filesDir/$v1WalletDirName/${BuildConfig.NetType}"
            val v1WalletExists = Dcrlibwallet.walletExistsAt(v1WalletPath)

            if (v1WalletExists) {

                setText(getString(R.string.migrating_wallet))

                MigrateV1Wallet(this, v1WalletPath) {
                    proceedToHomeActivity()
                }.beginV1WalletMigration()

            } else {
                launchSetupWalletActivity()
            }

        } else {
            checkStartupPass()
        }
    }

    private fun launchSetupWalletActivity() = GlobalScope.launch(Dispatchers.Default) {
        delay(3000)
        val i = Intent(this@SplashScreen, SetupWalletActivity::class.java)
        startActivity(i)
        finish()
    }

    private fun checkStartupPass() {
        if (multiWallet!!.isStartupSecuritySet) {
            requestStartupPass()
        } else {
            openWallet("")
        }
    }

    private fun requestStartupPass() {
        val title = PassPromptTitle(R.string.startup_password_prompt_title, R.string.startup_pin_prompt_title, R.string.startup_fingerprint_prompt_title)
        PassPromptUtil(this, null, title, allowFingerprint = true) { _, pass ->
            if (pass != null) {
                openWallet(pass)
            } else {
                endProcess()
            }

            true
        }.show()
    }

    private fun openWallet(publicPass: String) = GlobalScope.launch(Dispatchers.IO) {
        try {
            if (multiWallet!!.loadedWalletsCount() > 1) {
                setText(getString(R.string.opening_wallets))
            } else {
                setText(getString(R.string.opening_wallet))
            }

            multiWallet!!.openWallets(publicPass.toByteArray())

            proceedToHomeActivity()

        } catch (e: Exception) {
            e.printStackTrace()

            withContext(Dispatchers.Main) {
                val infoDialog = InfoDialog(this@SplashScreen)
                        .setDialogTitle(getString(R.string.failed_to_open_wallet))
                        .setMessage(Utils.translateError(this@SplashScreen, e))
                        .setPositiveButton(getString(R.string.exit_cap), DialogInterface.OnClickListener { _, _ -> endProcess() })

                if (e.message == Dcrlibwallet.ErrInvalidPassphrase) {

                    if (multiWallet!!.startupSecurityType() == Dcrlibwallet.PassphraseTypePin) {
                        infoDialog.setMessage(getString(R.string.invalid_pin))
                    }

                    infoDialog.setNegativeButton(getString(R.string.exit_cap), DialogInterface.OnClickListener { _, _ -> endProcess() })
                            .setPositiveButton(getString(R.string.retry_caps), DialogInterface.OnClickListener { _, _ ->
                                requestStartupPass()
                            })
                }

                infoDialog.setCancelable(false)
                infoDialog.setCanceledOnTouchOutside(false)
                infoDialog.show()
            }
        }
    }

    private fun proceedToHomeActivity() {
        val i = Intent(this@SplashScreen, HomeActivity::class.java)
                .addFlags(Intent.FLAG_ACTIVITY_CLEAR_TOP)
        startActivity(i)

        //Finish all the activities before this
        ActivityCompat.finishAffinity(this@SplashScreen)
    }

    private fun endProcess() {
        multiWallet?.shutdown()
        finish()
        exitProcess(1)
    }
}